runtime/runtime_events.c

```diff
@@ -277,10 +279,14 @@ static void runtime_events_create_from_stw_single(void) {
     );
 
     if (ring_file_handle == INVALID_HANDLE_VALUE) {
-      char* ring_loc_u8 = caml_stat_strdup_of_os(current_ring_loc);
-      caml_fatal_error("Couldn't open ring buffer loc: %s",
-                        ring_loc_u8);
-      caml_stat_free(ring_loc_u8);
+      char* ring_loc_u8 = caml_stat_strdup_noexc_os(current_ring_loc);
```

Comment:
The function `caml_stat_strdup_noexc_os` does a "os" -> "os" copy. Here you would need a "os" -> "utf-8" copy instead (like the original `caml_stat_strdup_of_os`).

Reply:
My mistake. I was tripped up by the similarity between the function names `caml_stat_strdup_of_os` (which returns a `char *` and can't raise an exception) and `caml_stat_strdup_os` (which returns a `char_os *` and can). I guess from the previous state of the code that I want MBCS for the `caml_fatal_error`, so I should use `caml_stat_strdup_of_os`. But my change was intended to avoid the (in fact impossible) exception, surviving a failure of the `strdup`. So looking at the implementation of `caml_stat_strdup_of_os`, it does a straight `malloc` under the hood, then calls `WideCharToMultiByte` with the un-validated return value. So that still seems unsafe: are we going to get NULL, or a SEGV?
This is getting quite bikesheddy: producing a useful `caml_fatal_error` message despite the failure of this particular `malloc` might not be a very high priority. But I can tweak `caml_stat_strdup_of_os`.

Reply:
> So looking at the implementation of `caml_stat_strdup_of_os`, it does a straight `malloc` under the hood, then calls `WideCharToMultiByte` with the un-validated return value.

Sorry, but `caml_stat_strdup_of_utf16` uses `caml_stat_alloc`, which (if memory serves) can raise "out of memory" exception; am I missing something?

https://github.com/ocaml/ocaml/blob/fc98dd14335964e371df8cda910295ba4c8ddf28/runtime/win32.c#L967-L977

Reply:
Argh, I'm touching the implementation in `yacc/wstr.c`, rather than the one in `runtime/win32.c`. PEBKAC. I'm going home.

Reply:
OK, I've cleaned this up with a new `caml_stat_strdup_noexc_of_os`.